src/QMCWaveFunctions/SPOSetBuilder.cpp

```diff
@@ -95,6 +95,10 @@ std::unique_ptr<SPOSet> SPOSetBuilder::createSPOSet(xmlNodePtr cur)
     app_error() << "Orbital optimization via rotation doesn't support complex wavefunction yet.\n";
     abort();
 #else
+    // At this point in the code we know that orbital rotation is used, so the target SPO
+    // can make a copy of the coefficient matrix.
+    sposet->storeParamsBeforeRotation();
```

Comment:
Can we move this line to the constructor of RoratedSPOs?

Reply:
Then that line needs to be guarded under a flag as in #4076 to distinguish between the original creation and the making of copies.  The point of moving storeParamsBeforeRotation here is to remove the need for the creatingCopy flag.

Reply:
Probably we need to make adjustment to the behavior makeClone, making sure it doesn't change the state of orbitals. Is this the reason that causes the threading bug?
@markdewing could you leave some comment about your understanding?
We can revisit the issue after this PR.

Reply:
The point of putting the storeParamsBeforeRotation here is then the SPOset is completely set up before any clones get made.  Then makeClone can make all the copies it wants and we don't have to worry about whether some of them are set up correctly (that is, have the pristine copy of the coefficient matrix initialized).

The other (so far unstated) requirement is that we know that the SPOSet is used for orbital rotation before making the copy of the coefficient matrix, in order to avoid wasting memory in cases that aren't going to use it.

Reply:
> The point of putting the storeParamsBeforeRotation here is then the SPOset is completely set up before any clones get made. 

If we move storeParamsBeforeRotation() to the constructor of RotatedSPO, it is still get called before any clones get made. However, if makeClone was invoking that constructor or doing something funny in the copy constructor, things get messed up.
About wasting memory, we need to change C_copy to shared_ptr like C. Since only RotatedSPO constructor calls storeParamsBeforeRotation. C_copy won't be created when rotation is not enabled.

Reply:
The RotatedSPOs::makeClone function does call the RotatedSPOs constructor.

Changing C_copy to shared_ptr will most likely cause the rotation to be applied <number of threads> times.  There will need to be some mechanism to ensure that the rotation only gets applied once.  This is something to address in the future.